File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:

        vis={}
        l,r=0,0
        count=0
        n=len(s)
        while r<n:
            vis[s[r]]=vis.get(s[r],0)+1
            r+=1
            while len(vis)==3 :
                count+= (n-(r)+1)
                vis[s[l]]-=1
                if vis[s[l]]==0:
                    del vis[s[l]]
                l+=1
            

        if len(vis)==3:
            count+=1
        return count



        

File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var numberOfSubstrings = function(s) {

    let vis={};
    let l=0;
    let r=0;

    let count=0;

    const n=s.length;
    while(r<n)
    {
        vis[s[r]]=(vis[s[r]] || 0) + 1;
        r++;
        while(Object.keys(vis).length===3)
        {
            count+=(n-r+1);
            vis[s[l]]--;
            if(vis[s[l]]===0)
            {
                delete vis[s[l]];
            }
            l++;
        } 
    }

    if(Object.keys(vis).length===3)
    {
        count++;
    }
    return count;
    
};